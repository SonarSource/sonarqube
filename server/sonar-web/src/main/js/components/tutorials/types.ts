/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export enum TutorialModes {
  Local = 'local',
  Jenkins = 'jenkins',
  BitbucketPipelines = 'bitbucket-pipelines',
  GitLabCI = 'gitlab-ci',
  GitHubActions = 'github-actions',
  AzurePipelines = 'azure-pipelines',
  OtherCI = 'other-ci',
}

export enum BuildTools {
  Maven = 'maven',
  Gradle = 'gradle',
  Cpp = 'cpp',
  ObjectiveC = 'objectivec',
  DotNet = 'dotnet',
  Dart = 'dart',
  Other = 'other',
}

export enum GradleBuildDSL {
  Groovy = 'build.gradle',
  Kotlin = 'build.gradle.kts',
}

export enum OSs {
  Linux = 'linux',
  Windows = 'win',
  MacOS = 'mac',
}

export enum Arch {
  X86_64 = 'x86_64',
  Arm64 = 'arm64',
}

export enum AutoConfig {
  Automatic = 'automatic',
  Manual = 'manual',
}

export type TutorialConfig = {
  autoConfig?: AutoConfig;
  buildTool?: BuildTools;
};
