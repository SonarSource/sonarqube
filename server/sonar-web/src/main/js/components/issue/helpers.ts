/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { IssueTransition } from '../../types/issues';

export function isTransitionDeprecated(transition: IssueTransition) {
  return transition === IssueTransition.Confirm || transition === IssueTransition.Resolve;
}

export function isTransitionHidden(transition: IssueTransition) {
  return transition === IssueTransition.WontFix;
}

export function transitionRequiresComment(transition: IssueTransition) {
  return [IssueTransition.Accept, IssueTransition.FalsePositive].includes(transition);
}
