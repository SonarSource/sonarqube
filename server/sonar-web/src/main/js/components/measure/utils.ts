/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { MetricKey } from '../../sonar-aligned/types/metrics';
import { Dict, Measure, MeasureEnhanced, MeasureIntern, Metric } from '../../types/types';

export const KNOWN_RATINGS = [
  MetricKey.sqale_rating,
  MetricKey.reliability_rating,
  MetricKey.security_rating,
  MetricKey.security_review_rating,
  MetricKey.software_quality_maintainability_rating,
  MetricKey.software_quality_reliability_rating,
  MetricKey.software_quality_security_rating,
  'maintainability_rating', // Needed to provide the label for "new_maintainability_rating"
];

export function enhanceMeasure(measure: Measure, metrics: Dict<Metric>): MeasureEnhanced {
  return {
    ...measure,
    metric: metrics[measure.metric],
    leak: getLeakValue(measure),
  };
}

export function getLeakValue(measure: MeasureIntern | undefined): string | undefined {
  return measure?.period?.value;
}

export function duplicationRatingConverter(val: number) {
  const value = val || 0;
  const THRESHOLD_B = 3;
  const THRESHOLD_C = 5;
  const THRESHOLD_D = 10;
  const THRESHOLD_E = 20;

  if (value === 0) {
    return 'A';
  } else if (value < THRESHOLD_B) {
    return 'B';
  } else if (value < THRESHOLD_C) {
    return 'C';
  } else if (value < THRESHOLD_D) {
    return 'D';
  } else if (value < THRESHOLD_E) {
    return 'E';
  }
  return 'F';
}
