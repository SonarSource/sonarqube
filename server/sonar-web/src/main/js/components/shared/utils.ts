/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { MetricKey } from '~sonar-aligned/types/metrics';
import { IssueStatus } from '../../types/issues';
import { Dict } from '../../types/types';

const ISSUE_MEASURES = [
  MetricKey.violations,
  MetricKey.new_violations,
  MetricKey.blocker_violations,
  MetricKey.critical_violations,
  MetricKey.major_violations,
  MetricKey.minor_violations,
  MetricKey.info_violations,
  MetricKey.new_blocker_violations,
  MetricKey.new_critical_violations,
  MetricKey.new_major_violations,
  MetricKey.new_minor_violations,
  MetricKey.new_info_violations,
  MetricKey.open_issues,
  MetricKey.reopened_issues,
  MetricKey.confirmed_issues,
  MetricKey.false_positive_issues,
  MetricKey.code_smells,
  MetricKey.new_code_smells,
  MetricKey.bugs,
  MetricKey.new_bugs,
  MetricKey.vulnerabilities,
  MetricKey.new_vulnerabilities,
  MetricKey.prioritized_rule_issues,
];

export const DEFAULT_ISSUES_QUERY = {
  issueStatuses: `${IssueStatus.Open},${IssueStatus.Confirmed}`,
};

const issueParamsPerMetric: Dict<Dict<string>> = {
  [MetricKey.blocker_violations]: { severities: 'BLOCKER' },
  [MetricKey.new_blocker_violations]: { severities: 'BLOCKER' },
  [MetricKey.critical_violations]: { severities: 'CRITICAL' },
  [MetricKey.new_critical_violations]: { severities: 'CRITICAL' },
  [MetricKey.major_violations]: { severities: 'MAJOR' },
  [MetricKey.new_major_violations]: { severities: 'MAJOR' },
  [MetricKey.minor_violations]: { severities: 'MINOR' },
  [MetricKey.new_minor_violations]: { severities: 'MINOR' },
  [MetricKey.info_violations]: { severities: 'INFO' },
  [MetricKey.new_info_violations]: { severities: 'INFO' },
  [MetricKey.open_issues]: { issueStatuses: IssueStatus.Open },
  [MetricKey.reopened_issues]: { issueStatuses: IssueStatus.Open },
  [MetricKey.confirmed_issues]: { issueStatuses: IssueStatus.Confirmed },
  [MetricKey.false_positive_issues]: { issueStatuses: IssueStatus.FalsePositive },
  [MetricKey.code_smells]: { types: 'CODE_SMELL' },
  [MetricKey.new_code_smells]: { types: 'CODE_SMELL' },
  [MetricKey.bugs]: { types: 'BUG' },
  [MetricKey.new_bugs]: { types: 'BUG' },
  [MetricKey.vulnerabilities]: { types: 'VULNERABILITY' },
  [MetricKey.new_vulnerabilities]: { types: 'VULNERABILITY' },
  [MetricKey.prioritized_rule_issues]: { prioritizedRule: 'true' },
};

export function isIssueMeasure(metric: string) {
  return ISSUE_MEASURES.indexOf(metric as MetricKey) !== -1;
}

export function propsToIssueParams(metric: string, inNewCodePeriod = false) {
  const params: Dict<string | boolean> = {
    ...DEFAULT_ISSUES_QUERY,
    ...issueParamsPerMetric[metric],
  };

  if (inNewCodePeriod) {
    params.inNewCodePeriod = true;
  }

  return params;
}
