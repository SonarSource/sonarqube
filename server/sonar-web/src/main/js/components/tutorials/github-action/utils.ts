/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export function generateGitHubActionsYaml(
  mainBranchName: string,
  branchesEnabled: boolean,
  runsOn: string,
  steps: string,
  additionalConfig?: string,
) {
  return `name: Build

on:
  push:
    branches:
      - ${mainBranchName}
${branchesEnabled ? '  pull_request:\n    types: [opened, synchronize, reopened]' : ''}

jobs:
  build:
    name: Build and analyze
    runs-on: ${runsOn}
    ${additionalConfig ?? ''}
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0  # Shallow clones should be disabled for a better relevancy of analysis${steps}`;
}
