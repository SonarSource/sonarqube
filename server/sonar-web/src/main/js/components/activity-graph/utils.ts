/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { chunk, flatMap, groupBy, sortBy } from 'lodash';
import { MetricKey, MetricType } from '~sonar-aligned/types/metrics';
import {
  CCT_SOFTWARE_QUALITY_METRICS,
  OLD_TO_NEW_TAXONOMY_METRICS_MAP,
} from '../../helpers/constants';
import { getLocalizedMetricName, translate } from '../../helpers/l10n';
import { localizeMetric } from '../../helpers/measures';
import { get, save } from '../../helpers/storage';
import { GraphType, MeasureHistory, ParsedAnalysis, Serie } from '../../types/project-activity';
import { Dict, Metric } from '../../types/types';

export const DEFAULT_GRAPH = GraphType.issues;

const GRAPHS_METRICS_DISPLAYED: Dict<string[]> = {
  [GraphType.issues]: [MetricKey.violations],
  [GraphType.coverage]: [MetricKey.lines_to_cover, MetricKey.uncovered_lines],
  [GraphType.duplications]: [MetricKey.ncloc, MetricKey.duplicated_lines],
};

const LEGACY_GRAPHS_METRICS_DISPLAYED: Dict<string[]> = {
  ...GRAPHS_METRICS_DISPLAYED,
  [GraphType.issues]: [MetricKey.bugs, MetricKey.code_smells, MetricKey.vulnerabilities],
};

const GRAPHS_METRICS: Dict<string[]> = {
  [GraphType.issues]: GRAPHS_METRICS_DISPLAYED[GraphType.issues].concat([
    MetricKey.reliability_rating,
    MetricKey.security_rating,
    MetricKey.sqale_rating,
  ]),
  [GraphType.coverage]: [...GRAPHS_METRICS_DISPLAYED[GraphType.coverage], MetricKey.coverage],
  [GraphType.duplications]: [
    ...GRAPHS_METRICS_DISPLAYED[GraphType.duplications],
    MetricKey.duplicated_lines_density,
  ],
};

const LEGACY_GRAPHS_METRICS: Dict<string[]> = {
  ...GRAPHS_METRICS,
  [GraphType.issues]: LEGACY_GRAPHS_METRICS_DISPLAYED[GraphType.issues].concat([
    MetricKey.reliability_rating,
    MetricKey.security_rating,
    MetricKey.sqale_rating,
  ]),
};

export const LINE_CHART_DASHES = [0, 3, 7];

export function isCustomGraph(graph: GraphType) {
  return graph === GraphType.custom;
}

export function getGraphTypes(ignoreCustom = false) {
  const graphs = [GraphType.issues, GraphType.coverage, GraphType.duplications];

  return ignoreCustom ? graphs : [...graphs, GraphType.custom];
}

export function hasDataValues(serie: Serie) {
  return serie.data.some((point) => Boolean(point.y || point.y === 0));
}

export function hasHistoryData(series: Serie[]) {
  return series.some((serie) => serie.data && serie.data.length > 1);
}

export function getSeriesMetricType(series: Serie[]) {
  return series.length > 0 ? series[0].type : MetricType.Integer;
}

export function getDisplayedHistoryMetrics(
  graph: GraphType,
  customMetrics: string[],
  isStandardMode = false,
) {
  if (isCustomGraph(graph)) {
    return customMetrics;
  }

  return isStandardMode ? LEGACY_GRAPHS_METRICS_DISPLAYED[graph] : GRAPHS_METRICS_DISPLAYED[graph];
}

export function getHistoryMetrics(
  graph: GraphType,
  customMetrics: string[],
  isStandardMode = false,
) {
  if (isCustomGraph(graph)) {
    return customMetrics;
  }
  return isStandardMode ? LEGACY_GRAPHS_METRICS[graph] : GRAPHS_METRICS[graph];
}

export function hasHistoryDataValue(series: Serie[]) {
  return series.some((serie) => serie.data && serie.data.length > 1 && hasDataValues(serie));
}

export function splitSeriesInGraphs(series: Serie[], maxGraph: number, maxSeries: number) {
  return flatMap(
    groupBy(series, (serie) => serie.type),
    (type) => chunk(type, maxSeries),
  ).slice(0, maxGraph);
}

export function generateCoveredLinesMetric(
  uncoveredLines: MeasureHistory,
  measuresHistory: MeasureHistory[],
): Serie {
  const linesToCover = measuresHistory.find(
    (measure) => measure.metric === MetricKey.lines_to_cover,
  );

  return {
    data: linesToCover
      ? uncoveredLines.history.map((analysis, idx) => ({
          x: analysis.date,
          y: Number(linesToCover.history[idx].value) - Number(analysis.value),
        }))
      : [],
    name: 'covered_lines',
    translatedName: translate('project_activity.custom_metric.covered_lines'),
    type: MetricType.Integer,
  };
}

export function generateSeries(
  measuresHistory: MeasureHistory[],
  graph: GraphType,
  metrics: Metric[],
  displayedMetrics: string[],
): Serie[] {
  if (displayedMetrics.length <= 0 || measuresHistory === undefined) {
    return [];
  }

  return sortBy(
    measuresHistory
      .filter((measure) => displayedMetrics.indexOf(measure.metric) >= 0)
      .map((measure) => {
        if (measure.metric === MetricKey.uncovered_lines && !isCustomGraph(graph)) {
          return generateCoveredLinesMetric(measure, measuresHistory);
        }
        const metric = findMetric(measure.metric, metrics);
        const isSoftwareQualityMetric = CCT_SOFTWARE_QUALITY_METRICS.includes(
          metric?.key as MetricKey,
        );
        return {
          data: measure.history.map((analysis) => {
            const { value } = analysis;

            return {
              x: analysis.date,
              y: metric?.type === MetricType.Level ? value : Number(value),
            };
          }),
          name: measure.metric,
          translatedName: metric ? getLocalizedMetricName(metric) : localizeMetric(measure.metric),
          type: !metric || isSoftwareQualityMetric ? MetricType.Integer : metric.type,
        };
      }),
    (serie) =>
      displayedMetrics.indexOf(
        serie.name === 'covered_lines' ? MetricKey.uncovered_lines : serie.name,
      ),
  );
}

export function saveActivityGraph(
  namespace: string,
  project: string,
  graph: GraphType,
  metrics?: string[],
) {
  save(namespace, graph, project);

  if (isCustomGraph(graph) && metrics) {
    save(`${namespace}.custom`, metrics.join(','), project);
  }
}

export function getActivityGraph(
  namespace: string,
  project: string,
): { customGraphs: string[]; graph: GraphType } {
  const customGraphs = get(`${namespace}.custom`, project);

  return {
    graph: (get(namespace, project) as GraphType) || DEFAULT_GRAPH,
    customGraphs: customGraphs ? customGraphs.split(',') : [],
  };
}

export function getAnalysisEventsForDate(analyses: ParsedAnalysis[], date?: Date) {
  if (date) {
    const analysis = analyses.find((a) => a.date.valueOf() === date.valueOf());
    if (analysis) {
      return analysis.events;
    }
  }

  return [];
}

export function getDeprecatedTranslationKeyForTooltip(metric: MetricKey) {
  const quality = OLD_TO_NEW_TAXONOMY_METRICS_MAP[metric];

  let deprecatedKey = 'severity';
  if (quality) {
    deprecatedKey = 'quality';
  } else if (metric === MetricKey.confirmed_issues) {
    deprecatedKey = 'confirmed';
  }

  return `project_activity.custom_metric.deprecated.${deprecatedKey}`;
}

function findMetric(key: string, metrics: Metric[]) {
  return metrics.find((metric) => metric.key === key);
}
