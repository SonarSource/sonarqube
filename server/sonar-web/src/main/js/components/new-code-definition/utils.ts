/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { hasGlobalPermission } from '../../helpers/users';
import {
  NewCodeDefinition,
  NewCodeDefinitionBranch,
  NewCodeDefinitionType,
} from '../../types/new-code-definition';
import { Permissions } from '../../types/permissions';
import { Component } from '../../types/types';
import { CurrentUser, isLoggedIn } from '../../types/users';

export enum NewCodeDefinitionLevels {
  Global = 'GLOBAL',
  Project = 'PROJECT',
  Branch = 'BRANCH',
  NewProject = 'NEW_PROJECT',
}

export type PreviouslyNonCompliantNCD = NewCodeDefinition &
  Required<Pick<NewCodeDefinition, 'previousNonCompliantValue' | 'updatedAt'>>;

export type PreviouslyNonCompliantBranchNCD = PreviouslyNonCompliantNCD & NewCodeDefinitionBranch;

export function isPreviouslyNonCompliantDaysNCD(
  newCodeDefinition: NewCodeDefinition,
): newCodeDefinition is PreviouslyNonCompliantNCD;
export function isPreviouslyNonCompliantDaysNCD(
  newCodeDefinition: NewCodeDefinitionBranch,
): newCodeDefinition is PreviouslyNonCompliantBranchNCD;
export function isPreviouslyNonCompliantDaysNCD(
  newCodeDefinition: NewCodeDefinition | NewCodeDefinitionBranch,
): newCodeDefinition is PreviouslyNonCompliantNCD | PreviouslyNonCompliantBranchNCD {
  return (
    newCodeDefinition.type === NewCodeDefinitionType.NumberOfDays &&
    newCodeDefinition.previousNonCompliantValue !== undefined &&
    newCodeDefinition.updatedAt !== undefined &&
    !newCodeDefinition.inherited
  );
}

export function isGlobalOrProjectAdmin(currentUser: CurrentUser, component?: Component) {
  if (!isLoggedIn(currentUser)) {
    return false;
  }

  if (component !== undefined) {
    return !!component.configuration?.showSettings;
  }

  return hasGlobalPermission(currentUser, Permissions.Admin);
}
