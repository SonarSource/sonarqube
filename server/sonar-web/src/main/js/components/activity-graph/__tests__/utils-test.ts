/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { MetricKey, MetricType } from '~sonar-aligned/types/metrics';
import * as dates from '../../../helpers/dates';
import { mockMeasureHistory, mockSerie } from '../../../helpers/mocks/project-activity';
import { get, save } from '../../../helpers/storage';
import { mockMetric } from '../../../helpers/testMocks';
import { GraphType } from '../../../types/project-activity';
import * as utils from '../utils';

jest.mock('date-fns', () => {
  const actual = jest.requireActual('date-fns');
  return {
    ...actual,
    startOfDay: jest.fn((date) => {
      const startDay = new Date(date);
      startDay.setUTCHours(0, 0, 0, 0);
      return startDay;
    }),
  };
});

jest.mock('../../../helpers/storage', () => ({
  save: jest.fn(),
  get: jest.fn(),
}));

const HISTORY = [
  mockMeasureHistory({
    metric: MetricKey.lines_to_cover,
    history: [
      { date: dates.parseDate('2017-04-27T08:21:32.000Z'), value: '100' },
      { date: dates.parseDate('2017-04-30T23:06:24.000Z'), value: '100' },
    ],
  }),
  mockMeasureHistory({
    metric: MetricKey.uncovered_lines,
    history: [
      { date: dates.parseDate('2017-04-27T08:21:32.000Z'), value: '12' },
      { date: dates.parseDate('2017-04-30T23:06:24.000Z'), value: '50' },
    ],
  }),
];

const METRICS = [
  mockMetric({ key: MetricKey.uncovered_lines, type: MetricType.Integer }),
  mockMetric({ key: MetricKey.lines_to_cover, type: MetricType.Percent }),
];

const SERIE = mockSerie({
  type: MetricType.Percent,
});

describe('generateCoveredLinesMetric', () => {
  it('should correctly generate covered lines metric', () => {
    expect(utils.generateCoveredLinesMetric(HISTORY[1], HISTORY)).toMatchSnapshot('with data');
    expect(utils.generateCoveredLinesMetric(HISTORY[1], [])).toMatchSnapshot('empty data');
  });
});

describe('generateSeries', () => {
  it('should correctly generate the series', () => {
    expect(
      utils.generateSeries(HISTORY, GraphType.coverage, METRICS, [
        MetricKey.uncovered_lines,
        MetricKey.lines_to_cover,
      ]),
    ).toMatchSnapshot();
  });
  it('should correctly handle non-existent data', () => {
    expect(utils.generateSeries(HISTORY, GraphType.coverage, METRICS, [])).toEqual([]);
  });
});

describe('getDisplayedHistoryMetrics', () => {
  const customMetrics = ['foo', 'bar'];
  it('should return only displayed metrics on the graph', () => {
    expect(utils.getDisplayedHistoryMetrics(utils.DEFAULT_GRAPH, [])).toEqual([
      MetricKey.violations,
    ]);
    expect(utils.getDisplayedHistoryMetrics(GraphType.coverage, customMetrics)).toEqual([
      MetricKey.lines_to_cover,
      MetricKey.uncovered_lines,
    ]);
  });
  it('should return all custom metrics for the custom graph', () => {
    expect(utils.getDisplayedHistoryMetrics(GraphType.custom, customMetrics)).toEqual(
      customMetrics,
    );
  });
  it('should return Legacy graphs', () => {
    expect(utils.getDisplayedHistoryMetrics(GraphType.issues, [], true)).toEqual([
      MetricKey.bugs,
      MetricKey.code_smells,
      MetricKey.vulnerabilities,
    ]);
  });
});

describe('getHistoryMetrics', () => {
  const customMetrics = ['foo', 'bar'];
  it('should return all metrics', () => {
    expect(utils.getHistoryMetrics(utils.DEFAULT_GRAPH, [])).toEqual([
      MetricKey.violations,
      MetricKey.reliability_rating,
      MetricKey.security_rating,
      MetricKey.sqale_rating,
    ]);
    expect(utils.getHistoryMetrics(GraphType.coverage, customMetrics)).toEqual([
      MetricKey.lines_to_cover,
      MetricKey.uncovered_lines,
      GraphType.coverage,
    ]);
    expect(utils.getHistoryMetrics(GraphType.custom, customMetrics)).toEqual(customMetrics);
  });
  it('should return legacy metrics', () => {
    expect(utils.getHistoryMetrics(utils.DEFAULT_GRAPH, [], true)).toEqual([
      MetricKey.bugs,
      MetricKey.code_smells,
      MetricKey.vulnerabilities,
      MetricKey.reliability_rating,
      MetricKey.security_rating,
      MetricKey.sqale_rating,
    ]);
  });
});

describe('hasHistoryData', () => {
  it('should correctly detect if there is history data', () => {
    expect(utils.hasHistoryData([mockSerie()])).toBe(true);
    expect(
      utils.hasHistoryData([
        mockSerie({
          data: [],
        }),
        mockSerie({
          name: 'bar',
          translatedName: 'bar',
        }),
      ]),
    ).toBe(true);
    expect(
      utils.hasHistoryData([
        mockSerie({
          name: 'bar',
          translatedName: 'bar',
          data: [{ x: dates.parseDate('2017-04-27T08:21:32.000Z'), y: 2 }],
        }),
      ]),
    ).toBe(false);
  });
});

describe('getGraphTypes', () => {
  it('should correctly return the graph types', () => {
    expect(utils.getGraphTypes()).toMatchSnapshot();
    expect(utils.getGraphTypes(true)).toMatchSnapshot();
  });
});

describe('hasDataValues', () => {
  it('should check for data value', () => {
    expect(utils.hasDataValues(SERIE)).toBe(true);
    expect(utils.hasDataValues({ ...SERIE, data: [] })).toBe(false);
  });
});

describe('getSeriesMetricType', () => {
  it('should return the correct type', () => {
    expect(utils.getSeriesMetricType([SERIE])).toBe(MetricType.Percent);
    expect(utils.getSeriesMetricType([])).toBe(MetricType.Integer);
  });
});

describe('hasHistoryDataValue', () => {
  it('should return the correct type', () => {
    expect(utils.hasHistoryDataValue([SERIE])).toBe(true);
    expect(utils.hasHistoryDataValue([])).toBe(false);
  });
});

describe('saveActivityGraph', () => {
  it('should correctly store data for standard graph types', () => {
    utils.saveActivityGraph('foo', 'bar', GraphType.issues);
    expect(save).toHaveBeenCalledWith('foo', GraphType.issues, 'bar');
  });

  it.each([[[]], [[MetricKey.bugs, MetricKey.alert_status]]])(
    'should correctly store data for custom graph types',
    (metrics) => {
      utils.saveActivityGraph('foo', 'bar', GraphType.custom, metrics);
      expect(save).toHaveBeenCalledWith('foo', GraphType.custom, 'bar');
      // eslint-disable-next-line jest/no-conditional-in-test
      expect(save).toHaveBeenCalledWith('foo.custom', metrics.join(','), 'bar');
    },
  );
});

describe('getActivityGraph', () => {
  it('should correctly retrieve data for standard graph types', () => {
    jest.mocked(get).mockImplementation((key) => {
      // eslint-disable-next-line jest/no-conditional-in-test
      if (key.includes('.custom')) {
        return null;
      }
      return GraphType.coverage;
    });

    expect(utils.getActivityGraph('foo', 'bar')).toEqual({
      graph: GraphType.coverage,
      customGraphs: [],
    });
  });

  it.each([null, 'bugs,code_smells'])(
    'should correctly retrieve data for custom graph types',
    (data) => {
      jest.mocked(get).mockImplementation((key) => {
        // eslint-disable-next-line jest/no-conditional-in-test
        if (key.includes('.custom')) {
          return data;
        }
        return GraphType.custom;
      });

      expect(utils.getActivityGraph('foo', 'bar')).toEqual({
        graph: GraphType.custom,
        // eslint-disable-next-line jest/no-conditional-in-test
        customGraphs: data ? [MetricKey.bugs, MetricKey.code_smells] : [],
      });
    },
  );

  it('should correctly retrieve data for unknown graphs', () => {
    jest.mocked(get).mockReturnValue(null);

    expect(utils.getActivityGraph('foo', 'bar')).toEqual({
      graph: GraphType.issues,
      customGraphs: [],
    });
  });
});
