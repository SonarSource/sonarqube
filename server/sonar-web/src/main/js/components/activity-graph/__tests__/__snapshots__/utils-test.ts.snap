// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCoveredLinesMetric should correctly generate covered lines metric: empty data 1`] = `
{
  "data": [],
  "name": "covered_lines",
  "translatedName": "project_activity.custom_metric.covered_lines",
  "type": "INT",
}
`;

exports[`generateCoveredLinesMetric should correctly generate covered lines metric: with data 1`] = `
{
  "data": [
    {
      "x": 2017-04-27T08:21:32.000Z,
      "y": 88,
    },
    {
      "x": 2017-04-30T23:06:24.000Z,
      "y": 50,
    },
  ],
  "name": "covered_lines",
  "translatedName": "project_activity.custom_metric.covered_lines",
  "type": "INT",
}
`;

exports[`generateSeries should correctly generate the series 1`] = `
[
  {
    "data": [
      {
        "x": 2017-04-27T08:21:32.000Z,
        "y": 88,
      },
      {
        "x": 2017-04-30T23:06:24.000Z,
        "y": 50,
      },
    ],
    "name": "covered_lines",
    "translatedName": "project_activity.custom_metric.covered_lines",
    "type": "INT",
  },
  {
    "data": [
      {
        "x": 2017-04-27T08:21:32.000Z,
        "y": 100,
      },
      {
        "x": 2017-04-30T23:06:24.000Z,
        "y": 100,
      },
    ],
    "name": "lines_to_cover",
    "translatedName": "lines_to_cover",
    "type": "PERCENT",
  },
]
`;

exports[`getGraphTypes should correctly return the graph types 1`] = `
[
  "issues",
  "coverage",
  "duplications",
  "custom",
]
`;

exports[`getGraphTypes should correctly return the graph types 2`] = `
[
  "issues",
  "coverage",
  "duplications",
]
`;
